exchanges/subscription/list.go

```diff
@@ -0,0 +1,38 @@
+package subscription
+
+import (
+	"slices"
+)
+
+// List is a container of subscription pointers
+type List []*Subscription
+
+// Strings returns a sorted slice of subscriptions
+func (l List) Strings() []string {
+	s := make([]string, len(l))
+	for i := range l {
+		s[i] = l[i].String()
+	}
+	slices.Sort(s)
+	return s
+}
+
+// GroupPairs groups subscriptions which are identical apart from the Pairs
+// The returned List contains cloned Subscriptions, and the original Subscriptions are left alone
+func (l List) GroupPairs() (n List) {
+	src := slices.Clone(l)
+	for len(src) > 0 {
+		s := src[0].Clone()
+		key := &IgnoringPairsKey{s}
+		n = append(n, s)
+		// Remove the first element, and any which match it
+		src = slices.DeleteFunc(src[1:], func(eachSub *Subscription) bool {
+			if key.Match(&IgnoringPairsKey{eachSub}) {
+				s.AddPairs(eachSub.Pairs...)
+				return true
+			}
+			return false
+		})
+	}
+	return
```

Comment:
Thanks 🙏 Though actually this matches one of my first implementations.
I moved back from it to the admittedly more complicated solution to try and get toa a single pass of the slice.
Specifically: We don't want to loop all of `n` for all of `l` to work out each time if this is the same.
So removing the elements was a deliberate choice.

Reply:
They both seem to benchmark similar with more pairs and look to be O(n2) here's another suggestion with benchmarks:

```go
type ignoreKey struct {
	Channel  string
	Asset    asset.Item
	Levels   int
	Interval kline.Interval
}

// GroupPairs groups subscriptions which are identical apart from the Pairs
// The returned List contains cloned Subscriptions, and the original Subscriptions are left alone
func (l List) GroupPairs() (n List) {
	lookup := make(map[ignoreKey]*Subscription)
	for _, src := range l {
		k := ignoreKey{Channel: src.Channel, Asset: src.Asset, Levels: src.Levels, Interval: src.Interval}
		pew, ok := lookup[k]
		if !ok {
			lookup[k] = src.Clone()
			continue
		}
		pew.Pairs = append(pew.Pairs, src.Pairs...)
	}
	n = make(List, len(l))
	i := 0
	for _, v := range lookup {
		n[i] = v
	}
	return
}
```

benchmark:

```go
// 30549	     40234 ns/op	   35850 B/op	     319 allocs/op (original)

// 36475	     33402 ns/op	   33441 B/op	      19 allocs/op (suggestion)

// 81321	     14935 ns/op	   33440 B/op	      19 allocs/op (map optimisation)
func BenchmarkGroupPairs(b *testing.B) {
	var pairs currency.Pairs
	for x := 0; x < 100; x++ {
		pairs = append(pairs, currency.NewPair(currency.NewCode(fmt.Sprintf("pair%d", x)), currency.USDT))
	}

	l := List{}
	for _, c := range []string{TickerChannel, OrderbookChannel} {
		for _, p := range pairs {
			l = append(l, &Subscription{
				Channel: c,
				Asset:   asset.Spot,
				Pairs:   currency.Pairs{p},
			})
		}
	}

	for i := 0; i < b.N; i++ {
		l.GroupPairs()
	}
}
```

Reply:
Nevermind with the map suggestion I accidently did a blind append for pairs and its actually a performance hit with the map. 
 //  32329	     35882 ns/op	   31666 B/op	      19 allocs/op

But the original suggestion seems like it beats out.

Reply:
Oh *so* please don't do that! 😂
You've bypassed ignoreKey and Store mechanics.
But I think the same solution works if we use a Store.
I think it won't be as performant, but we can't break responsibility encapsulation of both store and ignoreKey.

On the bench, I think you've reversed the O concern I have to focus on the Pairs.
I'm concerned about N *subs*, because that's where the double mapping would come into play.
I'll run some quick tests. We might be flogging a dead donkey on a relatively overdue PR though.

Reply:
Fixed 1544568

Added solution 4 to the mix, using store:
```
	s := NewStore()
	for _, sub := range l {
		if found := s.match(&IgnoringPairsKey{src}); found == nil {
			s.add(sub.Clone())
		} else {
			found.AddPairs(sub.Pairs...)
		}
	}
	return s.List()
```

The original is definitely surprisingly unperformant.
Adjusted Solution 3 to use `pew.AddPairs()` because duplicate pairs isn't acceptable.

```
BenchmarkGroupPairs-8                 86          13,700,565 ns/op          5,730,116 B/op    506,046 allocs/op
BenchmarkGroupPairs2-8               330           3,597,241 ns/op         1,670,692 B/op        962 allocs/op
BenchmarkGroupPairs3-8               650           1,826,868 ns/op         1,601,447 B/op        941 allocs/op
BenchmarkGroupPairs4-8               196           6,082,023 ns/op         1,682,938 B/op     11,116 allocs/op
```

Solution 4 is the best balance. It uses the safety of store and ignoreKeys without duplicating them, and it's fine in terms of performance for the heat of this path.

I actually feel really stupid for not thinking of (4) earlier.

Benchmark looked like:
```
func BenchmarkGroupPairs(b *testing.B) {
	channels := []string{}
	for x := 0; x < 100; x++ {
		channels = append(channels, fmt.Sprintf("ticker%d", x))
	}
	var pairs currency.Pairs
	for x := 0; x < 100; x++ {
		pairs = append(pairs, currency.NewPair(currency.NewCode(fmt.Sprintf("pair%d", x)), currency.USDT))
	}
	l := List{}
	for _, c := range channels {
		for _, p := range pairs {
			l = append(l, &Subscription{
				Channel: c,
				Asset:   asset.Spot,
				Pairs:   currency.Pairs{p},
			})
		}
	}
	for i := 0; i < b.N; i++ {
		l.GroupPairs()
	}
}
```